syntax = "proto3";

message VersionRequest { }

message VersionResponse{
    string version = 1;
}

message OpenDeviceRequest {
    string dev_name = 1;
}

message OpenDeviceResponce {
    string dev_name = 1;
}

message CreatePdRequest {
    string dev_name = 1;
}

message CreatePdResponse {
    uint32 pd_id = 1;
}

message CreateMrRequest {
    uint32 pd_id = 1;
    uint64 flag = 2;
    uint64 len = 3;
}

message CreateMrResponse {
    uint64 addr = 1;
    uint64 len = 2;
    uint32 rkey = 3;
    uint32 lkey = 4;
    uint32 mr_id = 5;
}

message CreateCqRequest {
    string dev_name = 1;
    uint32 cq_size = 2;
}

message CreateCqResponse {
    uint32 cq_id = 1;
}

message CreateQpRequest {
    uint32 pd_id = 1;
    enum QpType {
        RC = 0;
        UC = 1;
    }
    QpType qp_type = 2;
    uint32 cq_id = 3;
}

message CreateQpResponse {
    uint32 qp_id = 1;
    uint32 qp_num = 2;
}

message QueryPortRequest {
    string dev_name = 1;
    uint32 ib_port_num = 2;
}

message QueryPortResponse {
    uint32 lid = 1;
}

message QueryGidRequest {
    string dev_name = 1;
    uint32 ib_port_num = 2;
    uint32 gid_idx = 3;
}

message QueryGidResponse {
    bytes gid_raw = 1;
}

message ConnectQpRequest {
    string dev_name = 1;
    uint32 qp_id = 2;
    uint64 access_flag = 4;

    uint32 gid_idx = 8;
    uint32 ib_port_num = 7;
    uint32 remote_qp_num = 3;
    uint32 remote_lid = 5;
    bytes remote_gid = 6;

    uint32 timeout = 9;
    uint32 retry = 10;
    uint32 rnr_retry = 11;
}

message ConnectQpResponse {
}

message RemoteReadRequest {
    uint64 addr = 1;
    uint32 len = 2;
    uint32 lkey = 3;
    uint64 remote_addr = 4;
    uint32 remote_key = 5;
    uint32 qp_id = 6;
    uint32 cq_id = 7;
}

message RemoteReadResponse {
}

message LocalWriteRequest {
    uint32 mr_id = 1;
    uint32 offset = 2;
    uint32 len = 3;
    bytes content = 4;
}

message LocalWriteResponse {
}

message UnblockRetryRequest {
}

message UnblockRetryResponse {
}

message LocalCheckMemRequest {
    uint32 mr_id = 1;
    uint32 offset = 2;
    uint32 len = 3;
    bytes expected = 4;
}

message LocalCheckMemResponse {
    bool same = 1;
}

message RecvPktRequest {
    bool wait_for_retry = 1;
    bool has_cqe = 2;
    uint32 qp_id = 3;
}

message RecvPktResponse {}

message LocalRecvRequest {
    uint64 addr = 1;
    uint32 len = 2;
    uint32 lkey = 3;
    uint32 qp_id = 6;
    uint32 cq_id = 7;
}

message LocalRecvResponse {}

message RemoteSendRequest {
    uint64 addr = 1;
    uint32 len = 2;
    uint32 lkey = 3;
    uint32 qp_id = 6;
    uint32 cq_id = 7;
}

message RemoteSendResponse {}

message RemoteWriteRequest {
    uint64 addr = 1;
    uint32 len = 2;
    uint32 lkey = 3;
    uint64 remote_addr = 4;
    uint32 remote_key = 5;
    uint32 qp_id = 6;
    uint32 cq_id = 7;
}

message RemoteWriteResponse {
}